충남대 FE 정지원 - 2단계 상품 상세 페이지 API 구현하기

jjw5655 opened the pull request:
멘토님 안녕하세요! 2단계 구현입니다.

jjw5655 committed:
feat: 상품 상세 페이지 라우팅 및 기본 구조 구현

- 상품 상세 페이지 라우트 추가 (/products/:productId)
- ProductDetailPage 기본 컴포넌트 생성
- ProductDetailLayout 네비게이션 설정
- 상품 클릭 핸들러 수정 (주문 페이지 → 상품 상세 페이지)
- RankingSection, ThemeProductListPage에서 상품 클릭 시 상품 상세 페이지로 이동

jjw5655 committed:
feat: 상품 상세 페이지 API 구현

- 상품 상세 관련 타입 정의 (ProductDetail, ProductWish, ProductReview 등)
- 4개 API 함수 구현 (getProductDetail, getProductWish, getProductHighlightReview)
- React Query 훅 구현 (useProductDetail, useProductWish, useProductHighlightReview)
- ProductDetailPage에서 API 테스트 및 데이터 표시

jjw5655 committed:
feat: 상품 상세 페이지 UI 컴포넌트 구현

- ProductDetailHeader.tsx: 상품 헤더 컴포넌트 (이미지, 기본 정보)
- ProductDetailTabs.tsx: 탭 네비게이션 컴포넌트 (상품설명, 선물후기, 상세정보)
- ProductDetailContent.tsx: 탭 콘텐츠 컴포넌트 (상품 상세 정보, 리뷰 섹션)
- ProductDetailActions.tsx: 하단 액션 버튼 컴포넌트 (관심 등록 버튼, 주문하기 버튼)

jjw5655 committed:
feat: 상품 찜 기능 낙관적 업데이트 구현

- toggleWish API 함수 추가 (시뮬레이션)
- useWishMutation 훅 구현 (낙관적 업데이트, 에러 롤백)
- ProductDetailActions 컴포넌트에 mutation 연동
- 찜 버튼 클릭 시 즉시 UI 반영, API 호출 후 상태 유지
- 로딩 상태 처리 및 중복 클릭 방지

jjw5655 committed:
chore: TODO 주석 제거 (API 구현 완료)

jjw5655 committed:
feat: ErrorBoundary와 Suspense 적용으로 에러 처리 및 로딩 UI 개선

- ErrorBoundary 컴포넌트를 App.tsx와 ProductDetailPage에 적용하여 JavaScript 에러 캐치
- Suspense를 App.tsx에 적용하여 데이터 로딩 시 fallback UI 제공
- LoadingSkeleton 컴포넌트 생성 (text, image, button, card 타입 지원)
- 기존 단순 로딩 텍스트를 애니메이션 스켈레톤으로 개선

jjw5655 committed:
fix: 주문 페이지 뒤로가기 동작을 이전 페이지로 변경

- OrderLayout의 뒤로가기 버튼을 navigate(ROUTE_HOME)에서 navigate(-1)로 수정
- 상품 상세 페이지에서 주문 페이지로 이동 후 뒤로가기 시 상품 상세 페이지로 돌아가도록 함

jjw5655 committed:
docs: 미션 4 - 2단계 과제 요구사항으로 README.md 업데이트

terajh commented on src/App.tsx:
```diff
@@ -67,15 +85,16 @@ function App() {
+            onBackClick={handleBackClick}
+            onProfileClick={handleProfileClick}
+          />
+        )}
```
[의견] 지금 구현해주신 방향도 좋지만 location 이 ROUTE_ORDER 로 시작하지 않고, /products 로 시작하지 않을 때만 NavigationBar 를 노출시키고, 그 외에는 null을 노출하는 HOC 함수를 만들어 사용할 수 있을 것 같아 코멘트 남겨봅니다ㅎㅎ

jjw5655 replied:
별도의 유틸리티 함수로 분리했습니다!

terajh commented on src/api/product/mutations.ts:
```diff
@@ -0,0 +1,57 @@
+/**
+ * 상품 찜 토글을 위한 mutation 훅
+ */
+export const useWishMutation = () => {
```
찜을 뜻하는 wish 이군요! 네이밍 👍👍

jjw5655 replied:
👍👍👍

terajh commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -0,0 +1,102 @@
+    data: productReview,
+    isLoading: isLoadingReview,
+    error: errorReview,
+  } = useProductHighlightReview(productId);
```
[질문] 혹시 아래 캡쳐와 같이 react hook lint 에러 발생안하실까요?

<img width="670" height="628" alt="image" src="https://github.com/user-attachments/assets/cf7a407e-2d3d-4c59-8acb-847ba8a8b429" />

jjw5655 replied:
네 저한테는 아무 에러도 안뜨네요. npm install을 다시해봐도 에러가 뜨지 않고, 빌드도 문제없이 잘 되고 있습니다. 왜 거기서는 에러가 뜨는걸까요??

terajh replied:
린트가 제대로 동작하지 않는건가 싶네요ㅎㅎ 린트는 틈만나면 환경에 따라 에러가 발생하는 친구라 유심히 봐야하는것 같아요ㅋㅋ
이후 미션진행시 한번 node_modules/eslint/lib 에서 lint 룰이 정상적으로 불러와지는지 확인해보시는것 추천드립니다! 그래도 안된다면 다음 미션때 코멘트 부탁드려요

terajh commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -0,0 +1,102 @@
+        </div>
+      </Container>
+    );
+  }
```
현재는 에러 발생 시 직접 에러 컴포넌트를 렌더링하는 방식으로 개발하신 것 같습니다. 현재 상태도 충분히 좋지만, ErrorBoundary 컴포넌트를 구현해 활용하고, 에러를 throw하여 ErrorBoundary에서 에러 페이지를 구분해 보여주는 방식으로 리팩토링해 보는 것도 좋을 것 같습니다.

jjw5655 replied:
기존에 구현했던 ErrorBoundary와 ErrorFallback을 통해 처리하도록 수정했습니다. 현재는 에러의 종류가 2개(존재하지 않는 상품, API 오류)밖에 없어서 ErrorFallback에서는 에러 메시지만 구분하여 표시하도록 구현했습니다. 이후에 필요하다면 ErrorFallback에서 에러페이지를 더 구분하여 보여주도록 구현하겠습니다!

terajh commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -0,0 +1,102 @@
+        <LoadingSkeleton type="card" />
+      </Container>
+    );
+  }
```
TanStack Query를 사용하면 데이터 로딩 시 Suspense를 활용해 로딩 페이지를 쉽게 보여줄 수 있다는 장점이 있습니다. 따라서 기존 로딩 컴포넌트를 Suspense 기반으로 리팩토링해 보는 것도 좋을 것 같습니다.

jjw5655 replied:
리팩토링했습니다!

terajh reviewed (changes requested):
지원님~ 2단계 미션 구현 고생 많으셨습니다.

전반적으로 잘 구현해주신 것 같습니다. 커밋도 매우 세세하게 작성해주셔서 리뷰에 큰 도움이 되었습니다 👍👍 

다만, ErrorBoundary와 Suspense를 사용하려는 의도가 보이지만, 에러를 명시적으로 throw하는 로직이 없고, Suspense를 활용하지 않고 직접 로딩 컴포넌트를 반환하는 방식으로 구현하신 부분이 있었습니다. 기왕 ErrorBoundary와 Suspense를 사용하시는 김에, 에러 및 로딩 처리 로직을 리팩토링해 보시면 좋을 것같아 request changes 로 드려보도록 하겠습니다.

힌트를 드리자면 
> 로딩 처리는 각 컴포넌트에서 isLoading과 같은 상태나 props로 직접 로딩 UI를 보여주는 대신 Suspense를 활용하는 방식으로 변경해 보시고
> 에러 처리는 각 컴포넌트에서 isError로 직접 에러 UI를 렌더링하는 대신 throw Error를 통해 ErrorBoundary에서 에러를 캐치하고 에러 UI를 표시하도록 수정해 보시면 좋을 것 같습니다.
> useSuspenseQuery 를 살펴보셔도 좋을 것 같습니다 ㅎㅎ

과제를 진행하시면서 고민되었던 부분이나 궁금한 점이 있다면 PR description이나 DM으로 언제든 편하게 남겨주세요. 감사합니다! 🙏🏻

jjw5655 committed:
refactor: 네비게이션 표시 로직을 shouldShowNavigation 함수로 분리

jjw5655 committed:
refactor: 에러 처리 방식을 ErrorBoundary 활용으로 변경

- 에러 발생 시 throw
- ErrorFallback을 상품 정보 오류 전용으로 수정
- 데이터 없는 경우와 API 에러 상황으로 분리

jjw5655 committed:
refactor: TanStack Query Suspense 모드로 로딩 처리 리팩토링

- ProductDetailPage에서 수동 로딩 체크를 Suspense 기반으로 변경
  - TanStack Query v5는 개별 쿼리에서 suspense 옵션 설정 불필요
  - 데이터 로딩 시 자동으로 Suspense 동작하여 전역 처리 가능
- ProductDetailHeader, ProductDetailContent에서 isLoading props 제거
- 기존 LoadingSkeleton 컴포넌트 재활용으로 일관된 UX 제공

jjw5655 commented:
지금 다시보니까 코드를 진짜 대충 짜놨네요ㅋㅋ 사실 2단계 할 때 시간이 많이 없었습니다.

구현해 둔 ErrorBoundary나, suspense를 잘 활용하도록 리팩토링했습니다. 혹시 더 좋은 구현방식이나 제안이 있으면 많이 부탁드리겠습니다!

terajh commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -27,67 +26,37 @@ const ProductDetailPage = () => {
+  // Suspense 모드에서는 데이터가 준비되면 자동으로 렌더링됨
+  // productBasic이 null이면 API에서 해당 상품을 찾지 못한 것
+  if (productBasic === null) {
+    throw new Error(`존재하지 않는 상품입니다. (상품 ID: ${productId})`);
```
에러 타입별로 다른 형태의 에러핸들로직을 사용할 수 있게되었네요👍👍

terajh reviewed (approved):
코멘트 의견 반영해서 잘 리펙토링 해주셨군요! 고생하셨습니다. 

수정해주신 방법처럼 useQuery 의 api 요청이 실패한 경우 컴포넌트 레벨에서 error throw 하는 방법도 좋고, 그 외에는 tanstack query 자체에서 제공해주는 useSuspenseQuery 와 같은 suspense hook 을 활용해서도 쉽게 에러처리를 해줄 수 있을 것 같아요! 이후 미션 진행시 한번 공부해보시는것도 추천드립니다!

PR은 승인 후 머지하도록 하겠습니다. 고생하셨습니다~
